### DISCUSSION TITLE
Borrow React's ref disposing behavior

### DISCUSSION THREAD

**Discussion by astoilkov** [solidjs/solid#1558] [Ideas]
In React, when you do conditional rendering, the `ref` callback is called with `null` when the element is cleaned up. I propose the same behavior in Solid.

I think it makes sense because:
- Otherwise you are holding a value to an HTML element that's not in the DOM anymore.
- The behavior isn't consistent — initially the `ref` is `undefined` and then it's an HTML element.
- Feels more natural coming from React.

## Example

https://playground.solidjs.com/anonymous/a66ed951-70eb-4dbd-a6ec-26e19a33c010

```tsx
function App() {
  const [visible, setVisible] = createSignal(false)
  const [ref, setRef] = createSignal()

  createEffect(on(visible, () => {
    console.log(ref())
  }))

  return (
    <>
      <button onClick={() => setVisible(!visible())}>
        Toggle Visible
      </button>
      <Show when={visible()}>
        <span ref={setRef}>Visible</span>
      </Show>
    </>
  )
}
```

## Alternatives

I can think of one — to create an abstraction that will do this:

```ts
function createRef() {
  const [ref, setRef] = createSignal()

  return [ref, (element) => {
    setRef(element)
    onCleanup(() => {
      setRef(undefined)
    })
  }]
}
```

## Notes

- I feel calling the callback with `undefined` on cleanup instead of with `null` is more consistent because of the initial `undefined` value of the signal.
- This is a breaking behavior so, if decided, it can only be done in 2.0.

**Comment by ryansolid** [comment]
This was one of the big topics pre 1.0 release. And no surprise it is on my shortlist still. Funny enough the implementation would just be `onCleanup` like this internally. One observation was that sending in undefined/null caused a lot more guarding in the way code was written. Like simple examples got more complicated. Especially where ref cleaning wasn't necessary. Directives had a similar question. Should they call again without their element. Thinking about the way Solid manages lifecycles via primitives `onCleanup` is consistent. We don't re-render.

I'm not surprised the community at the time voted against it either. Very performance oriented. But things have progressed from then.

**Reply by astoilkov** [reply]
Yeah, TypeScript is probably the biggest problem here. Adding this removes convenience in the general case in order to allow an edge case — bummer.

You will figure it out.

For now, I've extracted a `createRef()` function (little hacky with returning a single function instead of two) that will require me to specify a type `createRef<HTMLDivElement | undefined>()` when the rendering is conditional. Type-safety breaks if you change the rendering but I will experiment with how much of a problem it is in the real world.

```ts
import { Accessor, createSignal, onCleanup } from 'solid-js'

export default function createRef<T>(): Accessor<T> {
  const [ref, setRef] = createSignal<T>()

  return (refValue?: T): T => {
    if (refValue !== undefined) {
      setRef(() => refValue)
      onCleanup(() => setRef(undefined))
    }

    return ref() as T
  }
}
```
